nomad/core_sched.go

```diff
@@ -880,13 +881,30 @@ func volumeClaimReapImpl(srv RPCServer, args *volumeClaimReapArgs) (map[string]i
 	// on the node need it, but we also only want to make this
 	// call at most once per node
 	if vol.ControllerRequired && args.nodeClaims[nodeID] < 1 {
+
+		// we need to get the CSI Node ID, which is not the same as
+		// the Nomad Node ID
+		ws := memdb.NewWatchSet()
+		targetNode, err := srv.State().NodeByID(ws, nodeID)
+		if err != nil {
+			return args.nodeClaims, err
+		}
+		if targetNode == nil {
+			return args.nodeClaims, fmt.Errorf("%s: %s",
+				structs.ErrUnknownNodePrefix, nodeID)
+		}
+		targetCSIInfo, ok := targetNode.CSINodePlugins[args.plug.ID]
+		if !ok {
+			return args.nodeClaims, fmt.Errorf("Failed to find NodeInfo for node: %s", targetNode.ID)
```

Comment:
It seems like this failure could be temporary, will we retry this again if the node plugin is in the process of coming back up? Also, is the node CSI id constant when the plugin comes back up or is it generated by the node plugin everytime it starts?

Reply:
The GC job will get requeued (which works fine for us with #7632 where the `Job.Deregister` will use the same GC path).

And as far as I can tell the CSI Node ID is implementation-specific. But in practice it has to be a fixed identifier of the underlying host. So for example, in the EBS case it's the EC2 instance ID, because if it wasn't the controller wouldn't be able to attach the external volume to it.